lib/translator/mode_sense.cc

```diff
@@ -0,0 +1,403 @@
+  header.dpofua = 0b01;
+  if (!cmd_attributes.dbd)
+    header.bdl = sizeof(scsi::ShortLbaBlockDescriptor);
+  else
+    header.bdl = 0;
+  if (!WriteValue(header, buffer)) {
+    DebugLog("Insufficient size for mode 6 parameter header");
+    return false;
+  }
+  return true;
+}
+
+// Writes a mode parameter 10 header to buffer
+bool WriteMode10Header(CommonCmdAttributes cmd_attributes,
+                       Span<uint8_t> buffer) {
+  scsi::ModeParameter10Header header = {};
+  header.mode_data_length = htons(GetModeDataLength(cmd_attributes, true));
+  header.dpofua = 0b01;
+  header.longlba = cmd_attributes.llbaa;
+  if (!cmd_attributes.dbd) {
+    if (cmd_attributes.llbaa)
+      header.bdl = htons(sizeof(scsi::LongLbaBlockDescriptor));
+    else
+      header.bdl = htons(sizeof(scsi::ShortLbaBlockDescriptor));
+  } else {
+    header.bdl = 0;
+  }
+  if (!WriteValue(header, buffer)) {
+    DebugLog("Insufficient size for mode 10 parameter header");
+    return false;
+  }
+  return true;
+}
+
+// Writes a device block descriptor to buffer
+// Section 6.3.2
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+bool WriteBlockDescriptor(const nvme::GenericQueueEntryCmd& identify,
+                          Span<uint8_t> buffer, bool llbaa,
+                          uint32_t& write_len) {
+  uint8_t* ns_dptr = reinterpret_cast<uint8_t*>(identify.dptr.prp.prp1);
+  Span<uint8_t> ns_span(ns_dptr, sizeof(nvme::IdentifyNamespace));
+  const nvme::IdentifyNamespace* idns;
+  idns = SafePointerCastRead<nvme::IdentifyNamespace>(ns_span);
+  if (llbaa) {
+    write_len = sizeof(scsi::LongLbaBlockDescriptor);
+    scsi::LongLbaBlockDescriptor lbd = {};
+    lbd.number_of_blocks = bswap_64(idns->ncap);
+    // Figure 246
+    // https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf
+    lbd.logical_block_length =
+        htonl(1 << (idns->lbaf[idns->flbas.format].lbads));
+    return WriteValue(lbd, buffer);
+  } else {
+    write_len = sizeof(scsi::ShortLbaBlockDescriptor);
+    scsi::ShortLbaBlockDescriptor sbd = {};
+    sbd.number_of_blocks = htonl(static_cast<uint32_t>(ltohll(idns->ncap)));
+    // Figure 246
+    // https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf
+    uint32_t lbl_be = htonl(1 << (idns->lbaf[idns->flbas.format].lbads));
+    if (IsLittleEndian()) lbl_be >>= 8;
+    sbd.logical_block_length = lbl_be;
+    return WriteValue(sbd, buffer);
+  }
+}
+
+// Writes corresponding mode page data to buffer
+void WritePageData(scsi::ModePageCode page_code, uint32_t get_features_result,
+                   Span<uint8_t> buffer) {
+  switch (page_code) {
+    case scsi::ModePageCode::kCacheMode: {
+      scsi::CachingModePage tmp_page = kCachingModePage;
+      // Figure 281 NVMe Spec
+      tmp_page.wce = 0b1 & get_features_result;
+      WriteValue(tmp_page, buffer);
```

Comment:
do you think it would be useful to have a check for these WriteValues

Reply:
Good point, I updated the code to support this